core/src/main/java/io/atomix/core/map/AsyncAtomicNavigableMap.java

```diff
@@ -0,0 +1,335 @@
+
+  /**
+   * Returns the least key greater than or equal to the given key,
+   * or {@code null} if there is no such key.
+   *
+   * @param key the key
+   * @return the least key greater than or equal to {@code key},
+   *         or {@code null} if there is no such key
+   * @throws ClassCastException if the specified key cannot be compared
+   *         with the keys currently in the map
+   * @throws NullPointerException if the specified key is null
+   *         and this map does not permit null keys
+   */
+  CompletableFuture<K> ceilingKey(K key);
+
+  /**
+   * Returns a key-value mapping associated with the least key
+   * strictly greater than the given key, or {@code null} if there
+   * is no such key.
+   *
+   * @param key the key
+   * @return an entry with the least key greater than {@code key},
+   *         or {@code null} if there is no such key
+   * @throws ClassCastException if the specified key cannot be compared
+   *         with the keys currently in the map
+   * @throws NullPointerException if the specified key is null
+   *         and this map does not permit null keys
+   */
+  CompletableFuture<Map.Entry<K, Versioned<V>>> higherEntry(K key);
+
+  /**
+   * Returns the least key strictly greater than the given key, or
+   * {@code null} if there is no such key.
+   *
+   * @param key the key
+   * @return the least key greater than {@code key},
+   *         or {@code null} if there is no such key
+   * @throws ClassCastException if the specified key cannot be compared
+   *         with the keys currently in the map
+   * @throws NullPointerException if the specified key is null
+   *         and this map does not permit null keys
+   */
+  CompletableFuture<K> higherKey(K key);
+
+  /**
+   * Returns a key-value mapping associated with the least
+   * key in this map, or {@code null} if the map is empty.
+   *
+   * @return an entry with the least key,
+   *         or {@code null} if this map is empty
+   */
+  CompletableFuture<Map.Entry<K, Versioned<V>>> firstEntry();
+
+  /**
+   * Returns a key-value mapping associated with the greatest
+   * key in this map, or {@code null} if the map is empty.
+   *
+   * @return an entry with the greatest key,
+   *         or {@code null} if this map is empty
+   */
+  CompletableFuture<Map.Entry<K, Versioned<V>>> lastEntry();
+
+  /**
+   * Removes and returns a key-value mapping associated with
+   * the least key in this map, or {@code null} if the map is empty.
+   *
+   * @return the removed first entry of this map,
+   *         or {@code null} if this map is empty
+   */
+  CompletableFuture<Map.Entry<K, Versioned<V>>> pollFirstEntry();
+
+  /**
+   * Removes and returns a key-value mapping associated with
+   * the greatest key in this map, or {@code null} if the map is empty.
+   *
+   * @return the removed last entry of this map,
+   *         or {@code null} if this map is empty
+   */
+  CompletableFuture<Map.Entry<K, Versioned<V>>> pollLastEntry();
```

Comment:
wasn't there problems with poll*?

Reply:
Yeah, it has to do with partitioned maps. Since `poll*` removes the first/last entry, running it against all the partitions will remove an entry per partition. Since we can only determine which was the lowest/highest entry _after_ it's already been removed, there's not really a way to do this atomically. But since ordered maps now have iterators, they have to be in a single partition now anyways. It's too difficult to maintain order across partitions for iterators.